core/services/job/spawner.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+//go:generate mockery --name Spawner --output ./mocks/ --case=underscore
+//go:generate mockery --name Delegate --output ./mocks/ --case=underscore
+
+type (
+	// The job spawner manages the spinning up and spinning down of the long-running
+	// services that perform the work described by job specs.  Each active job spec
+	// has 1 or more of these services associated with it.
+	//
+	// At present, Flux Monitor and Offchain Reporting jobs can only have a single
+	// "initiator", meaning that they only require a single service.  But the older
+	// "direct request" model allows for multiple initiators, which imply multiple
+	// services.
+	Spawner interface {
+		Start()
+		Stop()
+		CreateJob(spec Spec) (int32, error)
+		DeleteJob(ctx context.Context, spec Spec) error
+		RegisterDelegate(delegate Delegate)
+	}
+
+	spawner struct {
+		orm                    ORM
+		jobTypeDelegates       map[Type]Delegate
+		jobTypeDelegatesMu     sync.RWMutex
+		startUnclaimedServices utils.SleeperTask
+		services               map[int32][]Service
+		chStopJob              chan int32
+
+		utils.StartStopOnce
+		chStop chan struct{}
+		chDone chan struct{}
+	}
+
+	// TODO(spook): I can't wait for Go generics
+	Delegate interface {
+		JobType() Type
+		ToDBRow(spec Spec) models.JobSpecV2
+		FromDBRow(spec models.JobSpecV2) Spec
+		ServicesForSpec(spec Spec) ([]Service, error)
+	}
+)
+
+var _ Spawner = (*spawner)(nil)
+
+func NewSpawner(orm ORM) *spawner {
+	s := &spawner{
+		orm:              orm,
+		jobTypeDelegates: make(map[Type]Delegate),
+		services:         make(map[int32][]Service),
+		chStopJob:        make(chan int32),
+		chStop:           make(chan struct{}),
+		chDone:           make(chan struct{}),
+	}
+	s.startUnclaimedServices = utils.NewSleeperTask(
+		utils.SleeperTaskFuncWorker(s.startUnclaimedServicesWorker),
+	)
+	return s
+}
+
+func (js *spawner) Start() {
+	js.AssertNeverStarted()
+	go js.runLoop()
+}
+
+func (js *spawner) Stop() {
+	js.AssertNeverStopped()
+	close(js.chStop)
+	<-js.chDone
+}
+
+func (js *spawner) RegisterDelegate(delegate Delegate) {
+	js.jobTypeDelegatesMu.Lock()
+	defer js.jobTypeDelegatesMu.Unlock()
+
+	if _, exists := js.jobTypeDelegates[delegate.JobType()]; exists {
+		panic("registered job type " + string(delegate.JobType()) + " more than once")
+	}
+	logger.Infof("Registered job type '%v'", delegate.JobType())
+	js.jobTypeDelegates[delegate.JobType()] = delegate
+}
+
+func (js *spawner) runLoop() {
+	defer close(js.chDone)
+
+	// Initialize the Postgres event listener for new jobs
+	var chNewJobs <-chan string
+	listener, err := js.orm.ListenForNewJobs()
+	if err != nil {
+		logger.Errorw("Job spawner failed to subscribe to 'new job' events, falling back to polling", "error", err)
+	} else {
+		chNewJobs = listener.Events()
+	}
+
+	// Initialize the DB poll ticker
+	dbPollTicker := time.NewTicker(1 * time.Second)
+	defer dbPollTicker.Stop()
```

Comment:
Why?  I'd rather a simpler primitive that we don't have to `.Reset()` and just drop ticks that occur when the polling goroutine is already busy

Reply:
Does it drop ticks? I was under the impression that if it ticks while the goroutine is busy the tick will be queued and it will immediately enter the loop again, which is probably undesirable.